### PR TITLE
Add idealized marsh flooding case with vegetation drag

### PR DESCRIPTION
**Pull request by caozd999**
This PR is the sibling of #639 and a substitute of #552. 
In this PR, an idealized Delaware Bay coastal transect is defined as the combination of three parts from land to ocean: (1) coast (2) marsh land (where vegetation may occupy) and (3) offshore; each part is defined by its length and slope. Three test cases are set up by considering vegetation drag differently: (1) with no vegetation, and no vegetation drag; (2) with vegetation, represented by constant vegetation Manning's roughness (0.065, Dietrich et al., 2011); (3) with vegetation, represented by vegetation-induced Manning's roughness (equations of Moki et al., 2020). The test cases are driven by a semi-diurnal tidal boundary condition with tidal amplitude as 1.5 m. A new transect mesh is generated for this marsh flooding case.

**Results**
_(1) tidal amplitudes at the OB_
tidalcomparison

_(2) tidal water level along the transect_
movie_0810

**References**

> Moki, H., Taguchi, K., Nakagawa, Y., Montani, S. and Kuwae, T., 2020. Spatial and seasonal impacts of submerged aquatic vegetation (SAV) drag force on hydrodynamics in shallow waters. _Journal of Marine Systems_, p.103373.

> Dietrich, J.C., Westerink, J.J., Kennedy, A.B., Smith, J.M., Jensen, R.E., Zijlema, M., Holthuijsen, L.H., Dawson, C., Luettich Jr, R.A., Powell, M.D. and Cardone, V.J., 2011. Hurricane Gustav (2008) waves and storm surge: hindcast, synoptic analysis, and validation in Southern Louisiana. _Monthly Weather Review_, 139(8), pp.2488-2522.

### CODE REVIEW

**Comment by caozd999** [comment]
@sbrus89, please let me know if you have any comments. Thanks!

**Comment by sbrus89** [comment]
@caozd999, could you cherry-pick your last two commits onto ocean/coastal one more time. We had to do another rebase onto ocean/develop. After you do that, I'll test and merge.

**Comment by caozd999** [comment]
> @caozd999, could you cherry-pick your last two commits onto ocean/coastal one more time. We had to do another rebase onto ocean/develop. After you do that, I'll test and merge.

@sbrus89 sure I will work on it.

File: testing_and_setup/compass/ocean/drying_slope/forward_idealized_transect.xml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<template>
+	<namelist>
+		<option name="config_Rayleigh_friction">.false.</option>
+		<option name="config_Rayleigh_damping_depth_variable">.false.</option>
+		<option name="config_use_implicit_bottom_drag">.false.</option>
+		<option name="config_use_implicit_bottom_drag_variable">.false.</option>
+		<option name="config_use_implicit_bottom_drag_variable_mannings">.true.</option>
+		<option name="config_thickness_flux_type">'centered'</option>
+		<option name="config_use_idealized_transect">.true.</option>
```
Comment (sbrus89) [inline testing_and_setup/compass/ocean/drying_slope/forward_idealized_transect.xml:9]: `config_use_idealized_transect` is an init mode option right?

File: testing_and_setup/compass/ocean/drying_slope/init_idealized_transect.xml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<template>
+	<namelist>
+		<option name="config_idealized_transect_Lcoast">0.3</option>
```
Comment (sbrus89) [inline testing_and_setup/compass/ocean/drying_slope/init_idealized_transect.xml:3]: May need to set `config_use_idealized_transect` here?

Comment (caozd999) [reply testing_and_setup/compass/ocean/drying_slope/init_idealized_transect.xml:3]: Good point, thanks!

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/idealized_transect/config_init1.xml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+  <add_link source="../../../meshes/transect/build_transect_mesh/transectmesh.nc" dest="mesh.nc"/>
+  <add_link source="../../../meshes/transect/build_transect_mesh/graph.info" dest="graph.info"/>
+	<add_executable source="model" dest="ocean_model"/>
+
+	<namelist name="namelist.ocean" mode="init">
+		<template file="init_template.xml" path_base="script_configuration_dir"/>
+		<template file="init_idealized_transect.xml" path_base="script_configuration_dir"/>
+		<option name="config_use_variable_drag">.true.</option>
+		<option name="config_use_idealized_transect">.true.</option>
```
[the remark this answers is not in the document]
Comment (caozd999) [reply testing_and_setup/compass/ocean/drying_slope/marsh_flooding/idealized_transect/config_init1.xml:12]: I moved 
```
<option name="config_use_variable_drag">.true.</option>
<option name="config_use_idealized_transect">.true.</option>
```
to `init_idealized_transect.xml`

File: testing_and_setup/compass/ocean/drying_slope/forward_idealized_transect.xml (hunk 1 above)
Comment (caozd999) [reply testing_and_setup/compass/ocean/drying_slope/forward_idealized_transect.xml:9]: Yes, thanks for pointing it out.

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/idealized_transect/config_init1.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (sbrus89) [reply testing_and_setup/compass/ocean/drying_slope/marsh_flooding/idealized_transect/config_init1.xml:12]: Sounds good!

File: testing_and_setup/compass/ocean/drying_slope/meshes/transect/config_driver.xml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<driver_script name="run_test.py">
+	<case name="build_transect_mesh">
+		<step executable="./run.py" quiet="true" pre_message=" * Running build_transect_mesh" post_message="     Complete"/>
+	</case>
+</driver_script>
```
Comment (sbrus89) [inline testing_and_setup/compass/ocean/drying_slope/meshes/transect/config_driver.xml:5]: @caozd999, should the name of the directory for this step be `idealized_transect` instead of `transect` so that it matches with the corresponding forward case?

File: testing_and_setup/compass/ocean/drying_slope/init_idealized_transect.xml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+		<option name="config_idealized_transect_Sshore">0.0025</option>
+		<option name="config_idealized_transect_Lmarsh">0.5</option>
+		<option name="config_idealized_transect_Smarsh">0.0</option>
+		<option name="config_idealized_transect_roughness">0.025</option>
+		<option name="config_idealized_transect_roughness_marsh">0.025</option>
+		<option name="config_tidal_boundary_left_bottom_depth">0.0</option>
+		<option name="config_tidal_boundary_left_value">0.0</option>
+		<option name="config_tidal_boundary_right_value">8.0e3</option>
+		<option name="config_tidal_forcing_monochromatic_baseline">3.0</option>
+    <option name="config_alter_ICs_for_pbcs">.true.</option>
```
Comment (sbrus89) [inline testing_and_setup/compass/ocean/drying_slope/init_idealized_transect.xml:15]: This indent should be fixed

File: testing_and_setup/compass/ocean/drying_slope/meshes/transect/config_driver.xml (hunk 4 above)
Comment (caozd999) [reply testing_and_setup/compass/ocean/drying_slope/meshes/transect/config_driver.xml:5]: ok, I will change it accordingly. Thanks!

File: testing_and_setup/compass/ocean/drying_slope/init_idealized_transect.xml (hunk 5 above)
Comment (caozd999) [reply testing_and_setup/compass/ocean/drying_slope/init_idealized_transect.xml:15]: Fixed. Thanks for pointing it out.

Comment (caozd999) [reply testing_and_setup/compass/ocean/drying_slope/init_idealized_transect.xml]: Do you know how to fix indent issues more efficiently?

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/idealized_transect/config_forward1.xml [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+	<streams name="streams.ocean" keep="immutable" mode="forward">
+		<template file="forward_template.xml" path_base="script_configuration_dir"/>
+	</streams>
+
+	<run_script name="run.py">
+		<step executable="gpmetis">
+			<argument flag="graph.info">32</argument>
+		</step>
+		<model_run procs="32" threads="1" namelist="namelist.ocean" streams="streams.ocean"/>
```
Comment (sbrus89) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/idealized_transect/config_forward1.xml:20]: This could be changed to 36 so that it uses a full node on Grizzly.

Comment (caozd999) [reply testing_and_setup/compass/ocean/drying_slope/marsh_flooding/idealized_transect/config_forward1.xml]: fixed.

File: testing_and_setup/compass/ocean/drying_slope/meshes/transect/config_driver.xml (hunk 4 above)
Comment (sbrus89) [reply testing_and_setup/compass/ocean/drying_slope/meshes/transect/config_driver.xml]: Oh I see, that makes sense.

Comment (sbrus89) [reply testing_and_setup/compass/ocean/drying_slope/meshes/transect/config_driver.xml]: Feel free to change it back if you want

Comment (caozd999) [reply testing_and_setup/compass/ocean/drying_slope/meshes/transect/config_driver.xml]: Never mind, this looks good too -- it's now easier to link them.

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/idealized_transect/config_analysis.xml [hunk 7]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0"?>
+<config case="analysis">
+  <add_link source="../comparison.py" dest="comparison.py"/>
```
Comment (sbrus89) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/idealized_transect/config_analysis.xml:3]: @caozd999, there should be a `base_path` specified here for when cases are setup in different working directories

Comment (sbrus89) [reply testing_and_setup/compass/ocean/drying_slope/marsh_flooding/idealized_transect/config_analysis.xml:3]: I believe it should be:
`<add_link path_base="script_test_dir" source="comparison.py" dest="comparison.py"/>`

Comment (caozd999) [reply testing_and_setup/compass/ocean/drying_slope/marsh_flooding/idealized_transect/config_analysis.xml]: Thanks @sbrus89 , I have modified it.
